### PR TITLE
Update Merkle path convention

### PR DESCRIPTION
**Pull request by raychu86**
This PR updates how Merkle paths are represented, both natively and in its gadget form. We now only store the sibling hashes, instead of tuples of left /right hash. In addition to reducing constraint sizes, this change also allows us to do leaf update checks in the constraint world much more efficiently.

## Test Plan

The original tests should suffice for the merkle membership checks.

Additional tests for the leaf update functionality is a WIP.

### CODE REVIEW

File: algorithms/src/merkle_tree/merkle_path.rs [hunk 1]
```diff
@@ -70,17 +68,47 @@ impl<P: MerkleParameters> MerklePath<P> {
...
+        computed_hash: &<P::H as CRH>::Output,
+        sibling_hash: &<P::H as CRH>::Output,
+    ) -> Result<(<P::H as CRH>::Output, <P::H as CRH>::Output), MerkleError> {
+        let is_left = index & 1 == 0;
+        let mut left_bytes = computed_hash;
+        let mut right_bytes = sibling_hash;
+        if !is_left {
+            core::mem::swap(&mut left_bytes, &mut right_bytes);
+        }
+        Ok((left_bytes.clone(), right_bytes.clone()))
```
Comment (ljedrz) [inline algorithms/src/merkle_tree/merkle_path.rs:90]:
```suggestion
        Ok((*left_bytes, *right_bytes))
```

File: gadgets/src/algorithms/merkle_tree/merkle_path.rs [hunk 2]
```diff
@@ -137,27 +166,51 @@ where
...
     {
-        let mut path = Vec::new();
-        for (i, &(ref l, ref r)) in value_gen()?.borrow().path.iter().enumerate() {
-            let l_hash = HGadget::OutputGadget::alloc(&mut cs.ns(|| format!("l_child_{}", i)), || Ok(l.clone()))?;
-            let r_hash = HGadget::OutputGadget::alloc(&mut cs.ns(|| format!("r_child_{}", i)), || Ok(r.clone()))?;
-            path.push((l_hash, r_hash));
+        let merkle_path = value_gen()?.borrow().clone();
+
+        let pos_list: Vec<_> = merkle_path.position_list().collect();
+        let mut traversal = vec![];
```
Comment (ljedrz) [inline gadgets/src/algorithms/merkle_tree/merkle_path.rs:172]:
```suggestion
        let mut traversal = Vec::with_capacity(pos_list.len());
```

File: gadgets/src/algorithms/merkle_tree/merkle_path.rs [hunk 3]
```diff
@@ -137,27 +166,51 @@ where
...
+        for (i, position) in pos_list.iter().enumerate() {
+            traversal.push(Boolean::alloc(cs.ns(|| format!("alloc_position_{}", i)), || {
+                Ok(position)
+            })?);
         }
-        Ok(MerklePathGadget { path })
+
+        let mut path = vec![];
```
Comment (ljedrz) [inline gadgets/src/algorithms/merkle_tree/merkle_path.rs:179]:
```suggestion
        let mut path = Vec::with_capacity(merkle_path.path.len());
```

File: algorithms/src/merkle_tree/merkle_path.rs [hunk 4]
```diff
@@ -70,17 +68,47 @@ impl<P: MerkleParameters> MerklePath<P> {
...
+    }
+
+    /// The position of on_path node in `leaf_and_sibling_hash` and `non_leaf_and_sibling_hash_path`.
+    /// `position[i]` is 0 (false) iff `i`th on-path node from top to bottom is on the left.
+    ///
+    /// This function simply converts `self.leaf_index` to boolean array in big endian form.
+    pub fn position_list(&'_ self) -> impl '_ + Iterator<Item = bool> {
```
Comment (ljedrz) [inline algorithms/src/merkle_tree/merkle_path.rs:97]:
```suggestion
    pub fn position_list(&self) -> impl Iterator<Item = bool> + '_ {
```

File: gadgets/src/algorithms/merkle_tree/merkle_path.rs [hunk 5]
```diff
@@ -137,27 +166,51 @@ where
...
         T: Borrow<MerklePath<P>>,
     {
-        let mut path = Vec::new();
-        for (i, &(ref l, ref r)) in value_gen()?.borrow().path.iter().enumerate() {
-            let l_hash = HGadget::OutputGadget::alloc(&mut cs.ns(|| format!("l_child_{}", i)), || Ok(l.clone()))?;
-            let r_hash = HGadget::OutputGadget::alloc(&mut cs.ns(|| format!("r_child_{}", i)), || Ok(r.clone()))?;
-            path.push((l_hash, r_hash));
+        let merkle_path = value_gen()?.borrow().clone();
+
+        let pos_list: Vec<_> = merkle_path.position_list().collect();
```
Comment (ljedrz) [inline gadgets/src/algorithms/merkle_tree/merkle_path.rs:171]:
```suggestion
```

File: gadgets/src/algorithms/merkle_tree/merkle_path.rs [hunk 6]
```diff
@@ -137,27 +166,51 @@ where
...
+        let mut traversal = vec![];
+        for (i, position) in pos_list.iter().enumerate() {
```
Comment (ljedrz) [inline gadgets/src/algorithms/merkle_tree/merkle_path.rs:173]:
```suggestion
        for (i, position) in merkle_path.position_list().enumerate() {
```

File: gadgets/src/algorithms/merkle_tree/merkle_path.rs [hunk 7]
```diff
@@ -137,27 +166,51 @@ where
...
         T: Borrow<MerklePath<P>>,
     {
-        let mut path = Vec::new();
-        for (i, &(ref l, ref r)) in value_gen()?.borrow().path.iter().enumerate() {
-            let l_hash = HGadget::OutputGadget::alloc_input(&mut cs.ns(|| format!("l_child_{}", i)), || Ok(l.clone()))?;
-            let r_hash = HGadget::OutputGadget::alloc_input(&mut cs.ns(|| format!("r_child_{}", i)), || Ok(r.clone()))?;
-            path.push((l_hash, r_hash));
+        let merkle_path = value_gen()?.borrow().clone();
+
+        let pos_list: Vec<_> = merkle_path.position_list().collect();
```
Comment (ljedrz) [inline gadgets/src/algorithms/merkle_tree/merkle_path.rs:197]:
```suggestion
```

File: gadgets/src/algorithms/merkle_tree/merkle_path.rs [hunk 8]
```diff
@@ -137,27 +166,51 @@ where
...
+        let mut traversal = vec![];
+        for (i, position) in pos_list.iter().enumerate() {
```
Comment (ljedrz) [inline gadgets/src/algorithms/merkle_tree/merkle_path.rs:199]:
```suggestion
        for (i, position) in merkle_path.position_list().enumerate() {
```

File: gadgets/src/algorithms/merkle_tree/merkle_path.rs [hunk 9]
```diff
@@ -137,27 +166,51 @@ where
...
+            traversal.push(Boolean::alloc_input(
+                cs.ns(|| format!("alloc_input_position_{}", i)),
+                || Ok(position),
+            )?);
+        }
+
+        let mut path = vec![];
```
Comment (ljedrz) [inline gadgets/src/algorithms/merkle_tree/merkle_path.rs:206]:
```suggestion
        let mut path = Vec::with_capacity(merkle_path.path.len());
```

File: gadgets/src/algorithms/merkle_tree/tests.rs [hunk 10]
```diff
@@ -190,6 +190,63 @@ fn generate_masked_merkle_tree<P: MaskedMerkleParameters, F: PrimeField, HG: Mas
     assert!(cs.is_satisfied());
 }
 
+fn update_merkle_tree<P: MerkleParameters, F: PrimeField, HG: CRHGadget<P::H, F>>(leaves: &[[u8; 30]]) {
+    let parameters = P::default();
```
Comment (ljedrz) [inline gadgets/src/algorithms/merkle_tree/tests.rs:194]:
```suggestion
    let parameters = Arc::new(P::default());
```

File: gadgets/src/algorithms/merkle_tree/tests.rs [hunk 11]
```diff
@@ -190,6 +190,63 @@ fn generate_masked_merkle_tree<P: MaskedMerkleParameters, F: PrimeField, HG: Mas
...
+    let tree = MerkleTree::<P>::new(Arc::new(parameters.clone()), &leaves[..]).unwrap();
```
Comment (ljedrz) [inline gadgets/src/algorithms/merkle_tree/tests.rs:195]:
```suggestion
    let tree = MerkleTree::<P>::new(parameters.clone(), &leaves[..]).unwrap();
```

File: gadgets/src/algorithms/merkle_tree/tests.rs [hunk 12]
```diff
@@ -190,6 +190,63 @@ fn generate_masked_merkle_tree<P: MaskedMerkleParameters, F: PrimeField, HG: Mas
...
+
+    let mut satisfied = true;
+    for (i, leaf) in leaves.iter().enumerate() {
+        let proof = tree.generate_proof(i, &leaf).unwrap();
+        assert!(proof.verify(&root, &leaf).unwrap());
+
+        let mut updated_leaves = leaves.to_vec().clone();
+        updated_leaves[i] = [u8::MAX; 30];
+
+        let new_tree = MerkleTree::<P>::new(Arc::new(parameters.clone()), &updated_leaves[..]).unwrap();
```
Comment (ljedrz) [inline gadgets/src/algorithms/merkle_tree/tests.rs:206]:
```suggestion
        let new_tree = MerkleTree::<P>::new(parameters.clone(), &updated_leaves[..]).unwrap();
```

File: gadgets/src/algorithms/merkle_tree/tests.rs [hunk 13]
```diff
@@ -190,6 +190,63 @@ fn generate_masked_merkle_tree<P: MaskedMerkleParameters, F: PrimeField, HG: Mas
...
+    let parameters = P::default();
+    let tree = MerkleTree::<P>::new(Arc::new(parameters.clone()), &leaves[..]).unwrap();
+    let root = tree.root();
+
+    let mut satisfied = true;
+    for (i, leaf) in leaves.iter().enumerate() {
+        let proof = tree.generate_proof(i, &leaf).unwrap();
+        assert!(proof.verify(&root, &leaf).unwrap());
+
+        let mut updated_leaves = leaves.to_vec().clone();
```
Comment (ljedrz) [inline gadgets/src/algorithms/merkle_tree/tests.rs:203]:
```suggestion
        let mut updated_leaves = leaves.to_vec();
```

File: algorithms/src/merkle_tree/merkle_tree.rs [hunk 14]
```diff
@@ -322,10 +317,10 @@ impl<P: MerkleParameters + Send + Sync> MerkleTree<P> {
 
         if path.len() != Self::DEPTH as usize {
             let empty_hash = self.parameters.hash_empty()?;
-            path.push((self.tree[0].clone(), empty_hash));
+            path.push(empty_hash);
 
-            for &(ref hash, ref sibling_hash) in &self.padding_tree {
-                path.push((hash.clone(), sibling_hash.clone()));
+            for &(ref _hash, ref sibling_hash) in &self.padding_tree {
+                path.push(sibling_hash.clone());
```
Comment (ljedrz) [inline algorithms/src/merkle_tree/merkle_tree.rs:323]: If I'm not mistaken, the `padding_tree` is only ever populated with `(current_hash.clone(), empty_hash.clone())` tuples; doesn't this mean it is now redundant? Alternatively, doesn't it mean that the `sibling_hash` is always equal to `self.parameters.hash_empty()`?

Comment (howardwu) [reply algorithms/src/merkle_tree/merkle_tree.rs:323]: Based on this implementation, I think you're right.

Although given the padding tree is already computed by the time this function is called, I suppose it isn't significantly affecting performance.

I do think this padding approach is opinionated in its instantiation, though likely adequate for us.

**Review by howardwu — APPROVED** [review]

**Pull request merged**